nixos/onlyoffice: add allowLocalConnections option

bobberb opened the pull request:
Currently, Onlyoffice (which Nextcloud can use for the office-app suite) does not allow local connections. 
 
Resolves https://github.com/NixOS/nixpkgs/issues/317225
per [OnlyOffice Forums](https://forum.onlyoffice.com/t/how-to-allow-private-ip-to-access-onlyoffice-documentserver/5755)

<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [x] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [x] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

@SuperSandro2000

bobberb commented:
Seems nothing blocking this from release-24.05

SuperSandro2000 commented on nixos/modules/services/web-apps/onlyoffice.nix:
```diff
@@ -222,6 +228,13 @@ in
             cp -r ${cfg.package}/etc/onlyoffice/documentserver/* /run/onlyoffice/config/
             chmod u+w /run/onlyoffice/config/default.json
 
+            #Allow LAN-range IPs if set true
+            ${lib.optionalString (cfg.allowLocalConnections) ''
+            sed -i 's/"allowPrivateIPAddress": false/"allowPrivateIPAddress": true/' /run/onlyoffice/config/default.json
+            sed -i 's/"allowMetaIPAddress": false/"allowMetaIPAddress": true/' /run/onlyoffice/config/default.json
+            ''}
+
+
```
We don't want to use sed here if we have jq below.

bobberb replied:
Thanks for encouraging me to use the right tool! Done.

SuperSandro2000 commented on nixos/modules/services/web-apps/onlyoffice.nix:
```diff
@@ -222,6 +228,13 @@ in
+
             # Allow members of the onlyoffice group to serve files under /var/lib/onlyoffice/documentserver/App_Data
             chmod g+x /var/lib/onlyoffice/documentserver
 
```
I imagine we want to do the following:

- add a freeform json type option
- use pkgs.writeJSON on it
- merge it into the default.json in line 263

Can you do that? Thanks!

bobberb replied:
I do not think I have the knowledge to do a free-form JSON type but I did replace my commits with a new one that just uses the boolean and jq as you requested

SuperSandro2000 commented:
Please follow the [contributing guide](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#commit-conventions) when naming your commits.

bobberb commented:
@SuperSandro2000 will you accept this limited PR that does not introduce a free form json input (which I don't know how to do) but just adds the bool option into the existing jq structure?

raboof commented:
I guess this is fine to expose (though it'd be much neater to have a more general solution to #317225).

This particular PR now needs a rebase, reformat, and while the PR title was updated the commit title is still the old one.

I'm not a huge fan of the fact that the NixOS module option name doesn't directly map to the json option name, but I don't have a great suggestion either.

Commit:
nixos/onlyoffice: add allowLocalConnections option

Allow the document server to download files from private and
link-local IP addresses by setting `allowPrivateIPAddress` and
`allowMetaIPAddress` to true in the request-filtering-agent
configuration.

Resolves https://github.com/NixOS/nixpkgs/issues/317225

hatch01 reviewed (approved):
Thanks, @bobberb. LGTM.
This is a great first point to improve the usage of onlyoffice.
I may look into creating a freeform param to be able to configure everything if I have time.

raboof reviewed (approved):
> I may look into creating a freeform param to be able to configure everything if I have time.

That would be cool!
